src/mantarray_desktop_app/main_process/server.py

```diff
@@ -484,80 +479,26 @@ def set_protocols() -> Response:
             return Response(status=f"400 Protocol {protocol_id}, Invalid stimulation type: {stim_type}")
 
         # validate subprotocol dictionaries
-        for idx, subprotocol in enumerate(protocol["subprotocols"]):
-            subprotocol["type"] = subprotocol["type"].lower()
-            subprotocol_type = subprotocol["type"]
-            # validate subprotocol type
-            if subprotocol_type not in VALID_SUBPROTOCOL_TYPES:
-                return Response(
-                    status=f"400 Protocol {protocol_id}, Subprotocol {idx}, Invalid subprotocol type: {subprotocol_type}"
-                )
-
-            # validate subprotocol components
-            if subprotocol_type == "delay":
-                # make sure this value is not a float
-                subprotocol["duration"] = int(subprotocol["duration"])
-                total_subprotocol_duration_us = subprotocol["duration"]
-            else:  # monophasic and biphasic
-                max_abs_charge = (
-                    STIM_MAX_ABSOLUTE_VOLTAGE_MILLIVOLTS
-                    if stim_type == "V"
-                    else STIM_MAX_ABSOLUTE_CURRENT_MICROAMPS
-                )
-
-                subprotocol_component_validators = {
-                    "phase_one_duration": lambda n: n > 0,
-                    "phase_one_charge": lambda n: abs(n) <= max_abs_charge,
-                    "postphase_interval": lambda n: n >= 0,
-                }
-                if subprotocol_type == "biphasic":
-                    subprotocol_component_validators.update(
-                        {
-                            "phase_two_duration": lambda n: n > 0,
-                            "phase_two_charge": lambda n: abs(n) <= max_abs_charge,
-                            "interphase_interval": lambda n: n >= 0,
-                        }
-                    )
-                for component_name, validator in subprotocol_component_validators.items():
-                    if not validator(component_value := subprotocol[component_name]):  # type: ignore
-                        component_name = component_name.replace("_", " ")
-                        return Response(
-                            status=f"400 Protocol {protocol_id}, Subprotocol {idx}, Invalid {component_name}: {component_value}"
+        try:
+            for idx, subprotocol in enumerate(protocol["subprotocols"]):
```

Comment:
I think we should use recursion instead of looping in case we need to add more layers of nesting